### PR TITLE
[controller] do not delay RT topic deletion if the store is no longer hybrid

### PR DESCRIPTION
**Pull request by arjun4084346**
## Avoid `getStore` from the controller which is not the leader of cluster this store is a part of 

When TopicCleanupService tries to delete RT topics, it does not have to wait for all the corresponding VT topics to get deleted if all the versions of the store has changed from hybrid to batch. If all the versions are batch, it can be safely assumed that no VT topic is consuming from that RT topic.
This was already done in https://github.com/linkedin/venice/pull/1234 but that commit was reverted in https://github.com/linkedin/venice/pull/1355
It was reverted because TCS was trying to get `Store` for which it was not the leader controller. Only leader controller caches and stores the information for all the stores in the cluster it is leader of. 

So this time, `isRTTopicDeletionPermittedByAllControllers` will not take `Store` object, it will use the ControllerClientMap to get Store information for all the stores. 

Resolves #XXX

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java [hunk 1]
```diff
@@ -3551,24 +3554,37 @@ private void safeDeleteRTTopic(String clusterName, String storeName, Store store
...
+  }
+
+  private void deleteRealTtimeTopic(String clusterName, String storeName) {
+    Map<String, ControllerClient> controllerClientMap = getControllerClientMap(clusterName);
+    String rtTopicToDelete = Version.composeRealTimeTopic(storeName);
+    deleteRTTopicFromAllFabrics(rtTopicToDelete, controllerClientMap);
+    // Check if there is incremental push topic exist. If yes, delete it and send out to let other controller to
+    // delete it.
+    String incrementalPushRTTopicToDelete = Version.composeSeparateRealTimeTopic(storeName);
+    if (getTopicManager().containsTopic(pubSubTopicRepository.getTopic(incrementalPushRTTopicToDelete))) {
```
Comment (xunyin8) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java:3577]: Can we separate or extract the deletion of incremental push RT from RT? The triggers in this private method is a bit hard to follow and could be error prone (e.g. RT could be gone but incremental push RT is still around due to previous deletion failures).
1. The check for the existence of RT or need for RT deletion is outside of this private method.
2. The check for incremental push RT is only invoked in this private method.

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java [hunk 2]
```diff
@@ -3551,24 +3554,37 @@ private void safeDeleteRTTopic(String clusterName, String storeName, Store store
...
+      deleteRTTopicFromAllFabrics(incrementalPushRTTopicToDelete, controllerClientMap);
+    }
+  }
+
+  private void deleteRTTopicFromAllFabrics(String topic, Map<String, ControllerClient> controllerClientMap) {
```
Comment (xunyin8) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java:3582]: Is the goal here just trying to expedite the RT topic deletion in all fabrics?
The trigger for `deleteRTTopicFromAllFabrics` is only checking the local state. i.e. `!store.isHybrid() && getTopicManager().containsTopic(rtTopic)`.

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java [hunk 3]
```diff
@@ -3551,24 +3554,37 @@ private void safeDeleteRTTopic(String clusterName, String storeName, Store store
...
+    truncateKafkaTopic(topic);
```
Comment (xunyin8) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java:3583]: Did you verify if `controllerClientMap` excludes the local fabric? i.e. we might be making unnecessary duplicate calls to `truncateKafkaTopic`.

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java (hunk 2 above)
Comment (arjun4084346) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java:3582]: The goal is just to refactor this part and extract out method `isRTTopicDeletionPermittedByAllControllers` , so that `TopicCleanupService` can also call this method.

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java (hunk 3 above)
Comment (arjun4084346) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java:3583]: I don't think local fabric is excluded from `controllerClientMap`. removing the extra call

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java (hunk 1 above)
Comment (arjun4084346) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java:3577]: refactored.

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java (hunk 3 above)
Comment (arjun4084346) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java]: However, it was being done like that on left side too. I basically do not intend to change the code logic of this part, I am only trying to refactor this part to extract out `isRTTopicDeletionPermittedByAllControllers`

Comment (arjun4084346) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java:3578]: I put it back because a test `testHybridStoreToBatchOnly` was failing in `Assert.assertTrue(veniceAdmin.isTopicTruncated(rtTopic))`. It was not truncating the topic because `multiClusterConfigs.getControllerConfig(clusterName).getChildDataCenterControllerUrlMap()` was empty

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java [hunk 4]
```diff
@@ -3532,17 +3531,27 @@ private boolean hasFatalDataValidationError(PushMonitor pushMonitor, String topi
...
-    // we perform this check everytime when a store version is deleted we can afford to do best effort
+  private void safeDeleteRTTopic(String clusterName, String storeName) {
+    if (isRTTopicDeletionPermittedByAllControllers(clusterName, storeName)) {
+      String rtTopicToDelete = Version.composeRealTimeTopic(storeName);
+      deleteRTTopicFromAllFabrics(rtTopicToDelete, clusterName);
+    }
+    // Check if there is incremental push topic exist. If yes, delete it and send out to let other controller to
+    // delete it.
+    String incrementalPushRTTopicToDelete = Version.composeSeparateRealTimeTopic(storeName);
+    if (getTopicManager().containsTopic(pubSubTopicRepository.getTopic(incrementalPushRTTopicToDelete))) {
```
Comment (xunyin8) [inline services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java:3542]: The semantic of this refactor is actually different now. I think what you actually want here is save the result of `isRTTopicDeletionPermittedByAllControllers` to a boolean like `rtDeletionPermitted` and then here you want to have both condition to be true before deleting, i.e.
```
if (rtDeletionPermitted && getTopicManager().containsTopic(pubSubTopicRepository.getTopic(incrementalPushRTTopicToDelete)))
```
Otherwise it's not a safe delete

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java (hunk 1 above)
Comment (xunyin8) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java]: See new comment.

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java (hunk 2 above)
Comment (xunyin8) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java]: Ok, just wanted to ask because it's possible for each fabric to receive multiple requests to truncate topic. It should be harmless with the way we implemented the truncate API.

File: services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java (hunk 4 above)
Comment (arjun4084346) [reply services/venice-controller/src/main/java/com/linkedin/venice/controller/VeniceHelixAdmin.java:3542]: got it

**Review by xunyin8 — APPROVED** [review]
